R/Rules-methods.R

```diff
@@ -880,7 +882,157 @@ setMethod(
+#' @param SIM (`flag`)\cr is this method used in simulations? Default as `FALSE`.
+#'   If this flag is `TRUE` and target dose estimates (during trial and end-of-trial)
+#'   are outside of the dose grid range, the information message is printed by
+#'   this method.
+#'
+#' @aliases nextBest-NextBestMaxGain
+#'
+#' @export
+#' @example examples/Rules-method-nextBest-NextBestMaxGain.R
+#'
+setMethod(
+  f = "nextBest",
+  signature = signature(
+    nextBest = "NextBestMaxGain",
+    doselimit = "numeric",
+    samples = "missing",
+    model = "ModelTox",
+    data = "DataDual"
+  ),
+  definition = function(nextBest, doselimit, model, data, Effmodel, SIM = FALSE, ...) {
+    assert_class(Effmodel, "Effloglog")
+    assert_flag(SIM)
+    doselimit <- ifelse(missing(doselimit) || length(doselimit) == 0, Inf, doselimit)
+
+    # 'drt' - during the trial, 'eot' end of trial.
+    prob_target_drt <- nextBest@DLEDuringTrialtarget
+    prob_target_eot <- nextBest@DLEEndOfTrialtarget
+
+    # Target dose estimates, i.e. the dose with probability of the occurrence of
+    # a DLT that equals to the prob_target_drt or prob_target_eot.
+    dose_target_drt <- dose(x = prob_target_drt, model)
+    dose_target_eot <- dose(x = prob_target_eot, model)
+
+    # Find the dose which gives the maximum gain.
+    dose_grid_range <- h_dose_grid_range(data)
+    opt <- optim(
+      par = dose_grid_range[1],
+      fn = function(DOSE) {
+        -gain(DOSE, model_dle = model, model_eff = Effmodel)
+      },
+      method = "L-BFGS-B",
+      lower = dose_grid_range[1],
+      upper = dose_grid_range[2]
+    )
+    dose_mg <- opt$par # G*.
+    max_gain <- -opt$value
+
+    # Get eligible doses that do not exceed maximum possible dose.
+    is_dose_eligible <- data@doseGrid <= doselimit
+    # For placebo design, if safety allows, exclude placebo from the recommended next doses.
+    if (data@placebo && sum(is_dose_eligible) > 1L) {
+      is_dose_eligible <- is_dose_eligible[-1]
+    }
+    doses_eligible <- data@doseGrid[is_dose_eligible]
+
+    # Find the next best doses in the doseGrid. The next best dose is the dose
+    # at level closest and below the target. Here, there are four different
+    # targets: min(dose_mg, dose_target_drt), dose_mg, dose_target_drt or dose_target_eot.
+    # h_find_interval assumes that elements in doses_eligible are strictly increasing.
+    next_dose_lev <- h_find_interval(min(dose_mg, dose_target_drt), doses_eligible)
+    next_dose <- doses_eligible[next_dose_lev]
+
+    next_dose_mg_lev <- h_find_interval(dose_mg, doses_eligible)
+    next_dose_mg <- doses_eligible[next_dose_mg_lev]
+
+    next_dose_lev_drt <- h_find_interval(dose_target_drt, doses_eligible)
+    next_dose_drt <- doses_eligible[next_dose_lev_drt]
+
+    next_dose_lev_eot <- h_find_interval(dose_target_eot, doses_eligible)
+    next_dose_eot <- doses_eligible[next_dose_lev_eot]
+
+    # Find the variance of the log of dose_mg.
```

Comment:
I would suggest to factor this part out in helper function to reduce function length